pkg/api/v1alpha1/cloudstackmachineconfig_types.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// CloudStackMachineConfigSpec defines the desired state of CloudStackMachineConfig
+type CloudStackMachineConfigSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Template refers to a VM image template which has been previously registered in CloudStack. It can either be specified as a UUID or name
+	Template string `json:"template"`
+	// ComputeOffering refers to a compute offering which has been previously registered in CloudStack. It represents a VM’s instance size including number of CPU’s, memory, and CPU speed. It can either be specified as a UUID or name
+	ComputeOffering string `json:"computeOffering"`
+	// DiskOffering refers to a disk offering which has been previously registered in CloudStack. It represents a VM’s disk size and IOPS (Quality of Service) for primary data storage
+	DiskOffering string `json:"diskOffering,omitempty"`
+	// OSFamily is an EKS-A required field which represents the family of the guest OS image. Currently only “redhat“ is supported for CloudStack, which is also the default
+	OSFamily OSFamily `json:"osFamily,omitempty"`
+	// Details represents an arbitrary mapping of strings to strings, which are passed through to the CloudStack VM. Some example details which are commonly set on the VM include cpuOvercommitRatio, memoryOvercommitRatio, Message.ReservedCapacityFreed.Flag, and rootdisksize
+	Details map[string]string `json:"details,omitempty"`
```

Comment:
I could be missing something here but I would vote against this and in favor of something more structured. For example, if we want to allow users to configure cpuOvercommitRatio, memoryOvercommitRatio, Message.ReservedCapacityFreed.Flag, and rootdisksize, then create 4 more fields, one for each parameter.

This follows a bit better the patterns we have in the current API. It also reduces our testing surface since we control/restrict the number of possible configurations.

Reply:
I would be curious to hear what combinations we can expect a user to specify here. If there really are too many to track, I think since these details are supported as simple key/value pairs, I think it's okay. But if there are the main ones and we don't anticipate the user wanting fine-grain control of these fields, then I agree with creating new fields for structure.

Reply:
I'll ask some CloudStack experts and get back to you on this

Reply:
It sounds like the intention behind this field, even within CloudStack itself, is to support general-purpose arguments that needed to be added and available in the VM "object" in CloudStack's database. It's used as a catch-all API. I think it adds value in that if a customer wants to start using some new field specific to their use case, there's no code changes required in eks-a, or new release that has to be cut

Reply:
I get what you are saying
However, I would lean on the other side of the argument, which is that we want to control which fields a user can and can't set and that goes against the concept of a catch all API.

Part of the added value that EKS-A offers is the easiness of use and the reliability (achieved through extra validations and testing). And we get there by simplifying the CAPI and cloud provider APIs, adding our own abstraction on top of them. If we were gonna open this API to be able to set any kind of configuration, we would either have to:
* Exponentially increase our validations and testing matrix. I don't think we can realistically tackle this and even if we could, I don't think it will scale.
* Renounce to be able to cover and validate all the possible configuration. This would reduce the value of EKS-A.

That obviously doesn't mean the APIs have to be static. The idea is that we start with an opinionated proposal of what we think is best and iterate on it (adding, changing and removing) based on user's feedback and real case scenarios.

Reply:
We are gonna do more research about this and decide later if we need it or not
Tracking that work in #1187